lib/compress/zstd_compress_superblock.c

```diff
@@ -0,0 +1,611 @@
+                                                  sequences, nbSeq,
+                                                  llCode, mlCode, ofCode,
+                                                  cctxParams,
+                                                  op, oend-op,
+                                                  bmi2, writeEntropy);
+        FORWARD_IF_ERROR(cSeqSize);
+        if (cSeqSize == 0) return 0;
+        op += cSeqSize;
+    }
+    /* Write block header */
+    {   size_t cSize = (op-ostart)-ZSTD_blockHeaderSize;
+        U32 const cBlockHeader24 = lastBlock + (((U32)bt_compressed)<<1) + (U32)(cSize << 3);
+        MEM_writeLE24(ostart, cBlockHeader24);
+    }
+    return op-ostart;
+}
+
+/** ZSTD_compressSubBlock_multi() :
+ *  Breaks super-block into multiple sub-blocks and compresses them.
+ *  Entropy will be written to the first block.
+ *  The following blocks will use repeat mode to compress.
+ *  All sub-blocks are compressed blocks (no raw or rle blocks).
+ *  @return : compressed size of the super block (which is multiple ZSTD blocks)
+ *            Or 0 if it failed to compress. */
+static size_t ZSTD_compressSubBlock_multi(const seqStore_t* seqStorePtr,
+                            const ZSTD_entropyCTables_t* entropy,
+                            const ZSTD_entropyCTablesMetadata_t* entropyMetadata,
+                            const ZSTD_CCtx_params* cctxParams,
+                                  void* dst, size_t dstCapacity,
+                            const int bmi2, U32 lastBlock)
+{
+    const seqDef* const sstart = seqStorePtr->sequencesStart;
+    const seqDef* const send = seqStorePtr->sequences;
+    const seqDef* sp = sstart;
+    const BYTE* const lstart = seqStorePtr->litStart;
+    const BYTE* const lend = seqStorePtr->lit;
+    const BYTE* lp = lstart;
+    BYTE* const ostart = (BYTE*)dst;
+    BYTE* const oend = ostart + dstCapacity;
+    BYTE* op = ostart;
+    const BYTE* llCodePtr = seqStorePtr->llCode;
+    const BYTE* mlCodePtr = seqStorePtr->mlCode;
+    const BYTE* ofCodePtr = seqStorePtr->ofCode;
+    size_t targetCBlockSize = cctxParams->targetCBlockSize;
+    size_t litSize, seqCount;
+    int writeEntropy = 1;
+    size_t remaining = ZSTD_seqDecompressedSize(sstart, send-sstart, lend-lstart);
+    size_t cBlockSizeEstimate = 0;
+
+    DEBUGLOG(5, "ZSTD_compressSubBlock_multi (litSize=%u, nbSeq=%u)",
+                (unsigned)(lend-lp), (unsigned)(send-sstart));
+
+    litSize = 0;
+    seqCount = 0;
+    while (sp + seqCount < send) {
+        // TODO this is crude estimate for now...
+        // Ask Yann, Nick for feedback.
+        const seqDef* const sequence = sp + seqCount;
+        const U32 lastSequence = sequence+1 == send;
+        litSize = (sequence == send) ? (size_t)(lend-lp) : litSize + sequence->litLength;
+        seqCount++;
+        cBlockSizeEstimate = sizeof(seqDef) * seqCount + litSize;
```

Comment:
I haven't kept up too much with the discussion, but why do we want to break based on compressed size not uncompressed size? The latter we can easily compute.

Reply:
This is more in line with a transmission scenario : break data flow into small packets of roughly predictable size. Now, every ~`N` bytes, we have one block to decode, containing a variable number of bytes.

In contrast, splitting on uncompressed size, we end up with some very tiny blocks (big match) and others which are much larger (poorly compressible content) . It may also cut long matches in the middle.

Reply:
What would be the better way to estimate the compressed data size?

Reply:
CBlock Size estimation :
I see mostly 3 parts : 
1) Literals : starting from the huffman distribution presumed already calculated, the cost of each literal is known, in nb of bits. Just add these bits together. This requires reading the actual byte values from the literals buffer.
2) Sequences : same recipe, merely a bit more complex : the proportion of each symbol is presumed known, and from this information, it's possible to derive the cost of each symbol. There are several functions dedicated to this objective, at the beginning of`zstd_compress_sequences.c` .  The cost is supposed to be fractional (aka more precise that nb of bits). I think it uses a fixed point format.
3) Headers : Each block will cost a few additional bytes due to the format. Block size itself, but also size of literals section, nb of sequences, symbol type, etc. While this cost is not strictly speaking a "fixed cost", it's relatively stable, and can probably be cheaply estimated with a constant.

Reply:
Thanks for this! @Cyan4973 I have a question on #2. I am not sure if my understanding is correct but the functions to estimate cost of symbol only accounts for Code and does not account for additional bits for bitstream. Does those functions accounts for both Code and additional bits?

Reply:
You are totally right @ephiepark , the functions only estimate cost of the symbol. Additional bits must be added on top of that.